// Copyright (c) 2023 dingodb.com, Inc. All Rights Reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

import "coordinator.proto";
import "error.proto";

package dingodb.pb.push;

option java_package = "io.dingodb.coordinator";
option cc_generic_services = true;

message PushHeartbeatRequest {
  dingodb.pb.coordinator.StoreHeartbeatResponse heartbeat_response = 1;
}

message PushHeartbeatResponse {
  dingodb.pb.error.Error error = 1;
}

message RegionCmdResult {
  uint64 region_cmd_id = 1;
  dingodb.pb.error.Error error = 2;
}

message PushStoreOperationRequest {
  dingodb.pb.coordinator.StoreOperation store_operation = 1;
}

message PushStoreOperationResponse {
  dingodb.pb.error.Error error = 1;  // if this is OK, ignore region_cmd_results, else check every region_cmd_result
  repeated RegionCmdResult region_cmd_results = 2;
}

service PushService {
  // Hello
  rpc PushHeartbeat(PushHeartbeatRequest) returns (PushHeartbeatResponse);
  rpc PushStoreOperation(PushStoreOperationRequest) returns (PushStoreOperationResponse);
}
